### PR TITLE
MD/IB: added read of PCI info from files

### PR DESCRIPTION
**Pull request by hoopoepg**
- added reading of actual PCI BW from /sys/*** files

### CODE REVIEW

File: src/uct/ib/base/ib_md.c [hunk 1]
```diff
@@ -1411,6 +1451,58 @@ uct_ib_md_parse_subnet_prefix(const char *subnet_prefix_str,
...
+        return DBL_MAX; /* failed to read file */
+    } else {
+        pci_speed_str[len] = '\0';
+    }
+
+    if (sscanf(pci_width_str, "%u", &width) < 1) {
+        return DBL_MAX;
+    }
+
+    if (sscanf(pci_speed_str, "%lf", &bw) < 1) {
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: should we add GT/s to format string?

File: src/uct/ib/base/ib_md.c [hunk 2]
```diff
@@ -27,6 +27,16 @@
                                   IBV_ACCESS_REMOTE_ATOMIC)
 #define UCT_IB_MD_RCACHE_DEFAULT_ALIGN 16
 
+typedef struct uct_ib_md_pci_info {
+    double   bw;       /* bandwidth */
+    unsigned payload;  /* payload used to data transfer */
+    unsigned overhead; /* PHY + data link layer + header + *CRC* */
+    unsigned nack;     /* number of TLC before ACK */
+    unsigned ctrl;     /* length of control TLP */
+    unsigned encoding; /* number of bits in symbol encoded, 8 - gen 1/2, 128 - gen 3 */
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: could be float to be 8.0/10.0 or 128.0/130.0

File: src/uct/ib/base/ib_md.c [hunk 3]
```diff
@@ -151,6 +161,36 @@ static ucs_stats_class_t uct_ib_md_stats_class = {
 };
 #endif
 
+static uct_ib_md_pci_info_t uct_ib_md_pci_info[] = {
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: const?

File: src/uct/ib/base/ib_md.c [hunk 4]
```diff
@@ -27,6 +27,16 @@
                                   IBV_ACCESS_REMOTE_ATOMIC)
 #define UCT_IB_MD_RCACHE_DEFAULT_ALIGN 16
 
+typedef struct uct_ib_md_pci_info {
+    double   bw;       /* bandwidth */
+    unsigned payload;  /* payload used to data transfer */
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: could we make some fields uint16?

File: src/uct/ib/base/ib_md.c [hunk 5]
```diff
@@ -1411,6 +1451,58 @@ uct_ib_md_parse_subnet_prefix(const char *subnet_prefix_str,
...
 }
 
+static double uct_ib_md_read_pci_bw(struct ibv_device *ib_device)
+{
+    const char pci_data_path_fmt[]   = "/sys/class/infiniband/%s/device/%s";
+    const char pci_width_file_name[] = "current_link_width";
+    const char pci_speed_file_name[] = "current_link_speed";
+    char pci_width_str[16];
+    char pci_speed_str[16];
+    uct_ib_md_pci_info_t *p;
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: const ptr?

File: src/uct/ib/base/ib_md.c [hunk 6]
```diff
@@ -1411,6 +1451,58 @@ uct_ib_md_parse_subnet_prefix(const char *subnet_prefix_str,
     return UCS_OK;
 }
 
+static double uct_ib_md_read_pci_bw(struct ibv_device *ib_device)
+{
+    const char pci_data_path_fmt[]   = "/sys/class/infiniband/%s/device/%s";
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: could be const char *, to avoid stack usage

File: src/uct/ib/base/ib_md.c [hunk 7]
```diff
@@ -1411,6 +1451,58 @@ uct_ib_md_parse_subnet_prefix(const char *subnet_prefix_str,
...
+    uct_ib_md_pci_info_t *p;
+    double bw;
+    unsigned width;
+    ssize_t len;
+    size_t i;
+
+    len = ucs_read_file(pci_width_str, sizeof(pci_width_str) - 1, 1, pci_data_path_fmt,
+                        ib_device->name, pci_width_file_name);
+    if (len < 1) {
+        return DBL_MAX; /* failed to read file */
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: let's add ucs_debug for all error flows

File: src/uct/ib/base/ib_md.c [hunk 8]
```diff
@@ -1411,6 +1451,58 @@ uct_ib_md_parse_subnet_prefix(const char *subnet_prefix_str,
...
+    }
+
+    bw *= UCS_GBYTE / 8; /* to gigabit */
+
+    for (i = 0; i < ucs_static_array_size(uct_ib_md_pci_info); i++) {
+        if (bw < (uct_ib_md_pci_info[i].bw * 1.2)) { /* use 1.2 multiplex to avoid round issues */
+            p = &uct_ib_md_pci_info[i]; /* use pointer to make equation shorter */
+            return bw * width *
+                   (p->payload * p->nack) /
+                   (((p->payload + p->overhead) * p->nack) + p->ctrl) *
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: :+1:

Comment (yosefe) [reply src/uct/ib/base/ib_md.c]: let's add ucs_trace like:
`mlx5_0: pcie 8x gen3, effective throughput 13000 MB/s"
(13000 should be replaced by actual number)

File: src/uct/ib/base/ib_md.c [hunk 9]
```diff
@@ -1411,6 +1451,58 @@ uct_ib_md_parse_subnet_prefix(const char *subnet_prefix_str,
...
+
+    if (sscanf(pci_width_str, "%u", &width) < 1) {
+        return DBL_MAX;
+    }
+
+    if (sscanf(pci_speed_str, "%lf", &bw) < 1) {
+        return DBL_MAX;
+    }
+
+    bw *= UCS_GBYTE / 8; /* to gigabit */
```
Comment (yosefe) [inline src/uct/ib/base/ib_md.c]: you mean convert from Gbs to GBs - to giga**byte**

File: src/uct/ib/base/ib_md.c (hunk 2 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: float? there is number of bits used in transformation, how could it ba float? :)

File: src/uct/ib/base/ib_md.c [hunk 10]
```diff
@@ -1411,6 +1451,58 @@ uct_ib_md_parse_subnet_prefix(const char *subnet_prefix_str,
...
+    double bw;
+    unsigned width;
+    ssize_t len;
+    size_t i;
+
+    len = ucs_read_file(pci_width_str, sizeof(pci_width_str) - 1, 1, pci_data_path_fmt,
+                        ib_device->name, pci_width_file_name);
+    if (len < 1) {
+        return DBL_MAX; /* failed to read file */
+    } else {
```
Comment (brminich) [inline src/uct/ib/base/ib_md.c]: minor: you don't really need `else`, because you return in error case

File: src/uct/ib/base/ib_md.c (hunk 2 above)
Comment (yosefe) [reply src/uct/ib/base/ib_md.c]: i meant collapse encoding/decoding to single entry

Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: oh, i see
this is raw spec data (except overhead - there are tons of values), it is simpler to understand what PCI values are used having it in raw format

Comment (yosefe) [reply src/uct/ib/base/ib_md.c]: ok

File: src/uct/ib/base/ib_md.c (hunk 4 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: actually all fields (except bw) could be uint16. fixed

File: src/uct/ib/base/ib_md.c (hunk 3 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: yes, fixed

File: src/uct/ib/base/ib_md.c (hunk 6 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: hmmm, I expected that all const values are out of stack... but ok, fixed

File: src/uct/ib/base/ib_md.c (hunk 5 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: yes, fixed

File: src/uct/ib/base/ib_md.c (hunk 10 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: removed 'else'

File: src/uct/ib/base/ib_md.c (hunk 1 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: yep, why not

File: src/uct/ib/base/ib_md.c (hunk 9 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: right, comment is updated

File: src/uct/ib/base/ib_md.c (hunk 8 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: added trace MB/s and Gb/s (for bench and HCA speed)

File: src/uct/ib/base/ib_md.c (hunk 7 above)
Comment (hoopoepg) [reply src/uct/ib/base/ib_md.c]: added debug output

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8861/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8864/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8865/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8866/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8870/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
failed due to timeout

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8878/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Pull request merged**